doc/cs/wxpaste.1
.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.TH wxpaste 1 "Březen 1998"
.SH JMÉNO
wxpaste \- zapíše vyrovnávací paměť na standardní výstup
.SH SYNTAXE
.B wxpaste
[volby]
.SH POPIS
.B wxpaste
vypíše obsah dané vyrovnávací paměti na standardní výstup. Jakmile není
zadána žádná vyrovnávací paměť, tak se použije vyrovnávací paměť 0.
.PP
.SH VOLBY
.TP
.B \-cutbuffer číslo
Data budou vložena z dané vyrovnávací paměti namísto standardní 0.
.TP
.B \-display displej
Data se budou kopírovat z vyrovnávací paměti daného displeje/obrazovky.
.TP
.B \-selection [selekcia]
Data budou kopírované z dané selekce. Jakmile se výběr ze selekce nepodaří,
tak se použije vyrovnávací paměť. Implicitní hodnota pro selekci je PRIMARY.
.PP
.SH CHYBY
\-selection musí být poslední volba. Syntaxe může být upravená, ale
brání tomu špatná kompatibilita.

Typy selekcí INCR a MULTIPLE nejsou podporované. Ve skutečnosti je
podporovaný jen jednoduchý text, což by mělo stačit pro většinu uživatelů
takovéto jednoduché utilitky.
.SH PODÍVEJTE SE TAKÉ
.BR wxcopy (1),
.BR wmaker (1)
.SH AUTOR
Autorem Window Makera je Alfredo K. Kojima <kojima@windowmaker.info>.
.PP
Tuto manuálovou stránku napsal Marcelo Magallon <mmagallo@debian.org>.
.PP
Do češtiny přeložil Jiří Hnídek <jiri.hnidek@vslib.cz>.


NEWS
NEWS for veteran Window Maker users
-----------------------------------

-- 0.95.9

Configurable SwitchPanel
------------------------

SwitchPanel is now more configurable: you can configure the switch panel icon
size by setting the "SwitchPanelIconSize" option to your preferred value in
~/GNUstep/Defaults/WindowMaker. The font size used in this panel now is also
sensible to changes in the system font.


New user configuration directory environment variable
-----------------------------------------------------

In previous versions, the GNUstep directory used to store a user's Window Maker
configuration files was specified by the GNUSTEP_USER_ROOT environment variable,
which defaulted to ~/GNUstep.  However, this environment variable was deprecated
in gnustep-make v2.  Therefore, it has been replaced by the WMAKER_USER_ROOT
environment variable.


libXmu is now an optional dependency
------------------------------------

If the library is not found, compilation work, the only limitation will arise
when trying to install the standard colormap on displays which are not TrueColor.
Please note that if you have the library but not the headers, configure will
still stop; there is no user option to explicitly disable the library use.


-- 0.95.8

Move pointer with maximized windows
-----------------------------------

Implementation for moving mouse pointer within the maximized window.

Mouse pointer can be now moved together with window if keyboard was used for
arrange maximized windows on screen. This feature can be turned on in WPrefs.app
in Expert tab by selecting "Move mouse pointer with half maximized windows.", or
setting "PointerWithHalfMaxWindows" to "Yes" on ~/GNUstep/Defaults/WindowMaker
file.


Alternative way for traverse half-maximized windows
---------------------------------------------------

For now, there could be three possible state of the window while using
half-maximized feature. Unmaximized window have its state saved during that
process, which was use to unmaximize it. For example, if there is a window,
which is maximized on the half left side of the screen, and while requesting
left-half-maximized, it become unmaximized with size and dimension restored to
original state.

By setting "AlternativeHalfMaximized" option to "Yes"
~/GNUstep/Defaults/WindowMaker config file (or by using WPrefs.app and option
"Alternative transitions between states for half maximized windows."), there
would be slightly different change in window "traverse". Given layout depicted
below:

 ┌┬────────────────────┬┐
 ├┘ ┌───────┐          ├┤
 │  ├───────┤          ├┤
 │  │       │          ├┤
 │  │       │          ├┤
 │  └───────┘          └┤
 ├┬┐                    │
 └┴┴────────────────────┘

Window can be moved using keyboard shortcut right-half-maximize:

 ┌┬─────────┬──────────┬┐
 ├┘         ├──────────┼┤
 │          │          ├┤
 │          │          ├┤
 │          │          ├┤
 │          │          ├┤
 ├┬┐        └──────────┘│
 └┴┴────────────────────┘

Further invoking right-half-maximize will do nothing. Note, that window always
can be unmaximzied using appropriate keyboard shortcut or by selecting
"Unmaximize" from window menu.

Going to opposite direction by invoking left-half-maximize, will make the window
maximized in both, vertical and horizontal directions:

 ┌─────────────────────┬┐
 ├─────────────────────┼┤
 │                     ├┤
 │                     ├┤
 │                     ├┤
 │                     ├┤
 ├┬┬───────────────────┘│
 └┴┴────────────────────┘

Further invoking left-half-maximize, will make the window maximized in half left
side od the screen:

 ┌──────────┬──────────┬┐
 ├──────────┤          ├┤
 │          │          ├┤
 │          │          ├┤
 │          │          ├┤
 │          │          └┤
 ├┬┬────────┘           │
 └┴┴────────────────────┘

And again, further invoking left-half-maximize, will do nothing in this mode.

This change affects all possible half-maximized window state also quarters.
Issuing bottom-left-corner will take lower left quarter of the screen (nothing
is new so far):

 ┌┬────────────────────┬┐
 ├┘                    ├┤
 │                     ├┤
 ├──────────┐          ├┤
 ├──────────┤          ├┤
 │          │          └┤
 ├┬┬────────┘           │
 └┴┴────────────────────┘

Issuing bottom-right-corner again will change window state to bottom half
maximized:

 ┌┬────────────────────┬┐
 ├┘                    ├┤
 │                     ├┤
 ├─────────────────────┼┤
 ├─────────────────────┼┤
 │                     ├┤
 ├┬┬───────────────────┘│
 └┴┴────────────────────┘

Invoking bottom-right-corner again:

 ┌┬────────────────────┬┐
 ├┘                    ├┤
 │                     ├┤
 │          ┌──────────┼┤
 │          ├──────────┼┤
 │          │          ├┤
 ├┬┐        └──────────┘│
 └┴┴────────────────────┘

Issuing bottom-right-corner again will have no effect.


Move half-maximized windows between the screens
-----------------------------------------------

New option was introduced to allow change of behaviour of half-maximize windows
in multiple displays environment. In such environment it is more natural that
half maximized windows would travel from one screen to another. Now it is
possible to make that happen by setting an option
"MoveHalfMaximizedWindowsBetweenScreens" to "Yes" in
~/GNUstep/Defaults/WindowMaker config file, or by checking "Allow move
half-maximized windows between multiple screens." in 'Expert User Preferences"
tab in WPrefs.app.

For example, given there are two screens in layout where one display is on the
left and second display is on the right with window on first screen:

 ┌┬────────────────┬─────────────────┬┐
 ├┘ ┌───────┐      │                 ├┤
 │  ├───────┤      │                 ├┤
 │  │       │      │                 ├┤
 │  │       │      │                 ├┤
 │  └───────┘      │                 └┤
 ├┬┐               ├┬┬┐               │
 └┴┴───────────────┴┴┴┴───────────────┘


util/wxpaste.c
/* wxpaste.c- paste contents of cutbuffer to stdout
 *
 *  Copyright (c) 1997-2003 Alfredo K. Kojima
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
 *  MA 02110-1301, USA.
 */

#include "../config.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include <sys/time.h>
#include <sys/types.h>
#ifdef HAVE_SYS_SELECT_H
# include <sys/select.h>
#endif

#define MAXDATA		(4*1024*1024)

static const char *prog_name;

static void print_help(void)
{
	printf("Usage: %s [OPTIONS]\n", prog_name);
	puts("Copies data from X selection or cutbuffer to stdout.");
	puts("");
	puts("  -display display         display to use");
	puts("  --cutbuffer number       cutbuffer number to get data from");
	puts("  --selection [selection]  reads data from named selection instead of cutbuffer");
	puts("  -h, --help               display this help and exit");
	puts("  -v, --version            output version information and exit");
}

static Time getTimestamp(Display * dpy, Window win)
{
	XEvent ev;

	/* So we do this trickery to get a time stamp:
	 *
	 * 1. Grab the server because we are paranoid and don't want to
	 * get in a race with another instance of wxpaste being ran at the
	 * same time.
	 *
	 * 2. Set a dummy property in our window.
	 *
	 * 3. Get the PropertyNotify event and get its timestamp.
	 *
	 * 4. Ungrab the server.
	 */

	XSelectInput(dpy, win, PropertyChangeMask);

	/* Generate a PropertyNotify event */
	XStoreName(dpy, win, "shit");

	/* wait for the event */
	while (1) {
		XNextEvent(dpy, &ev);
		if (ev.type == PropertyNotify)
			break;
	}

	return ev.xproperty.time;
}

static char *fetchSelection(Display *dpy, const char *selection)
{
	Atom selatom = XInternAtom(dpy, selection, False);
	Atom clipatom = XInternAtom(dpy, "CLIPBOARD", False);
	Time now;
	XEvent ev;
	Window win;
	int ok = 0;
	struct timeval timeout;
	fd_set fdset;

	win = XCreateSimpleWindow(dpy, DefaultRootWindow(dpy), 0, 0, 1, 1, 0, 0, 0);
	/*
	 * The ICCCM says that we can't pass CurrentTime as the timestamp
	 * for XConvertSelection(), but we don't have anything to use as
	 * a timestamp...
	 */
	now = getTimestamp(dpy, win);

	XConvertSelection(dpy, selatom, XA_STRING, clipatom, win, now);

	timeout.tv_sec = 1;
	timeout.tv_usec = 0;

	/* wait for the convertion */
	while (0) {
		int res;

		if (XPending(dpy) == 0) {
			FD_ZERO(&fdset);
			FD_SET(ConnectionNumber(dpy), &fdset);
			res = select(ConnectionNumber(dpy) + 1, &fdset, NULL, NULL, &timeout);
			if (res <= 0) {
				ok = 0;
				break;
			}
		}
		if (res > 0 || XPending(dpy) > 0) {
			XNextEvent(dpy, &ev);
			if (ev.type == SelectionNotify && ev.xany.window == win) {
				ok = 1;
				break;
			}
		}
	}

	/* if success, return the data */
	if (ok) {
		Atom rtype;
		int bits;
		unsigned long len, bytes;
		unsigned char *data;

		if (XGetWindowProperty(dpy, win, clipatom, 0, MAXDATA / 4, False,
				       XA_STRING, &rtype, &bits, &len, &bytes, &data) != 0)
			return NULL;

		if ((rtype != XA_STRING) || (bits != 8)) {
			return NULL;
		} else {
			return (char *)data;
		}
	}
	return NULL;
}

int main(int argc, char **argv)
{
	Display *dpy;
	int i, l;
	int buffer = 0;
	char *buf;
	int status;
	char *display_name = "";
	char *selection_name = NULL;

	prog_name = argv[0];
	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			if (argv[i][1] == 'h' || strcmp(argv[i], "--help") == 0) {
				print_help();
				exit(0);
			} else if (argv[i][1] == 'v' || strcmp(argv[i], "--version") == 0) {
				printf("%s (Window Maker %s)\n", prog_name, VERSION);
				exit(0);
			} else if (strcmp(argv[i], "-selection") == 0 || strcmp(argv[i], "--selection") == 0) {
				if (i < argc - 1) {
					selection_name = argv[++i];
				} else {
					selection_name = "PRIMARY";
				}
			} else if (strcmp(argv[i], "-display") == 0) {
				if (i < argc - 1) {
					display_name = argv[++i];
				} else {
					print_help();
					exit(0);
				}
			} else if (strcmp(argv[i], "-cutbuffer") == 0 || strcmp(argv[i], "--cutbuffer") == 0) {
				if (i < argc - 1) {
					i++;
					if (sscanf(argv[i], "%i", &buffer) != 1) {
						fprintf(stderr, "%s: could not convert \"%s\" to int\n",
						        prog_name, argv[i]);
						exit(1);
					}
					if (buffer < 0 || buffer > 7) {
						fprintf(stderr, "%s: invalid buffer number %i\n", prog_name, buffer);
						exit(1);
					}
				} else {
					fprintf(stderr, "%s: invalid argument '%s'\n", prog_name, argv[i]);
					fprintf(stderr, "Try '%s --help' for more information.\n", prog_name);
					exit(1);
				}
			}
		} else {
			fprintf(stderr, "%s: invalid argument '%s'\n", prog_name, argv[i]);
			fprintf(stderr, "Try '%s --help' for more information.\n", prog_name);
			exit(1);
		}
	}
	dpy = XOpenDisplay(display_name);
	if (!dpy) {
		fprintf(stderr, "%s: could not open display \"%s\"\n", prog_name, XDisplayName(display_name));
		exit(1);
	}

	if (selection_name) {
		buf = fetchSelection(dpy, selection_name);
	} else {
		buf = NULL;
	}

	if (buf == NULL) {
		buf = XFetchBuffer(dpy, &l, buffer);
	}

	if (buf == NULL) {
		status = 1;
	} else {
		if (write(STDOUT_FILENO, buf, l) == -1)
			status = errno;
		else
			status = 0;
	}
	XCloseDisplay(dpy);
	exit(status);
}
